### ISSUE TITLE
mojo clients that live outside the mojo tree need a test embedder

### ISSUE THREAD

**Issue by jynnantonix** [domokit/mojo#766]
It's currently not possible to run any go tests for code that touches mojoms because the compiled test binary still needs something to inject the raw system thunks.  The mojo edk has a test embedder that's used by the go tests in //mojo/go:system_tests.  This should be exposed outside the edk in some way so that clients that don't live in the mojo tree can still test their code.

@vardhan-cr @shinykitten @viettrungluu-cr

**Comment by viettrungluu-cr** [comment]
I don't think we want to do this. People should be writing "app tests",
i.e., tests that are mojo apps.

**Comment by jynnantonix** [comment]
Let me try this again.  I have the following problem:

I am currently writing go code that lives outside the mojo tree and that implements some mojo interfaces.  I would like to test this code.  Go has very powerful and easy testing infrastructure.  For testing file foo.go, I create a file foo_test.go that imports the testing package and define my tests to be top-level functions with a "Test" prefix.  Running these tests is just running the command `go test` in the package directory.

The `go test` command parses all the *_test.go files in the package and then [generates a main package](https://github.com/golang/go/blob/master/src/cmd/go/test.go#L1423) that gets built into an executable.  Unfortunately, right now that executable segfaults as soon as it tries to create a message pipe because there is no embedder to inject the raw system thunks.

Even if you say that people should be writing "app tests" it's not clear what that means in this context.  Due to the way tests are built, it's not clear to me how I would even build a mojo app out of the tests in a package.  One way might be to basically re-implement the logic in the `go test` command and have it generate a main package that implements the mojo Application interface.  That seems like a lot of work and the internal API that the `go test` command uses is not guaranteed to be stable across releases of the language.  Maybe @crawshaw can comment on how feasible that is.

Another way might be to not use the go testing infrastructure at all and use a separate custom mechanism.  This would make me very unhappy because it would mean losing all the benefits of the go test infrastructure, which includes things like CPU and memory profiling, benchmarks, and code coverage.  And while I may be willing to work with this, it would also mean that some external developers would take one look at our framework and decide to use something else because it doesn't use the standard go conventions.  We might as well drop go support if we think we're going to go down this route.

So dismissing this issue by saying that people should be writing "app tests" doesn't actually help me in any way.  Either mojo needs to provide the appropriate hooks to play nicely with the existing go test infrastructure or mojo needs to provide a test embedder package that people can just import in their tests.  The latter seems much easier to do.

**Comment by jamesr** [comment]
It'd be useful to provide unit test fakes for independent pieces of logic that wish to interact with some mojo primitives but this would need to be done for higher level components, not the low level embedder API.  Tests that wish to interact with a real Mojo system are integration tests and won't work with the "go test" framework.

**Comment by jynnantonix** [comment]
So the specific case that I'm having trouble with is testing an implementation of a go interface that uses a mojo service under the hood.  See [this CL](https://fuchsia-review.googlesource.com/#/c/2802).  What I'd like to do is spin up a fake implementation of the mojo service and then connect that to the mojo client, which would be tested through the interface tests.  Right now this crashes because CreateMessagePipe is a null function pointer and needs to be set to a real function by an embedder.

**Comment by jamesr** [comment]
Right, for that use case I think what you should be able to do is provide a fake implementation of the mojo interface - i.e. an implementation of interface block.Device.  CreateMessagePipe is several layers down and not something your code depends on directly so it probably shouldn't be involved in the test at all.  The go bindings are concrete in a few places that make this hard to do today but I think that should be addressed at the bindings layer higher up.
